Deterministically order class members for emission

slavapestov opened the pull request:
We already did this for vtable entries, because its required for correctness; sink the ordering down into EmittedMembersRequest in service of reproducible builds as well.

This fixes an issue found by @davidungar where the driver's batching could flip the order of a class's default init() and deinit members, which tripped up our incremental build testing.

Along the way, I discovered that caching the emitted members list broke deriving the `Differentiable` conformance, because we might cache the member list before synthesizing the members; fixing this by forcing the conformance in `EmittedMembersRequest` in turn uncovered some order dependencies in conformance derivation, which I already addressed in https://github.com/apple/swift/pull/32578.

slavapestov committed:
Sema: Use request evaluator caching for EmittedMembersRequest

slavapestov committed:
Sema: Force synthesized conformance to Differentiable in EmittedMembersRequest

slavapestov committed:
Sema: Deterministically order declarations in EmittedMemberRequest

Previously we did this in the SILVTableVisitor, since maintaining a
consistent order between translation units is important for correctness.

However, we also want the order in which the members themselves are
emitted to remain consistent regardless of the order in which members
were synthesized. This is in service of reproducible builds.

slavapestov committed:
ASTMangler: Tiny cleanup

slavapestov committed:
AST: Mark implicit destructor as synthesized to fix non-deterministic function order

We can synthesize the default init() and the implicit deinit in a class in
any order, depending on the order of primary files and how that class was
used in other primary files.

Make sure that EmittedMembersRequest puts the destructor at the end with
all other synthesized members, so that we produce the same object file in
any case.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2451,16 +2503,40 @@ EmittedMembersRequest::evaluate(Evaluator &evaluator,
   forceConformance(Context.getProtocol(KnownProtocolKind::Decodable));
   forceConformance(Context.getProtocol(KnownProtocolKind::Encodable));
   forceConformance(Context.getProtocol(KnownProtocolKind::Hashable));
+  forceConformance(Context.getProtocol(KnownProtocolKind::Differentiable));
```
@dan-zheng @rxwei I think this should be uncontroversial. We need to force the derived members, if any, to be synthesized here, because the result of this request is now cached.

rxwei replied:
Looks good to me.

davidungar commented on lib/AST/Decl.cpp:
```diff
@@ -4318,6 +4318,10 @@ GetDestructorRequest::evaluate(Evaluator &evaluator, ClassDecl *CD) const {
     CD->recordObjCMethod(DD, DD->getObjCSelector());
 
+  // Mark it as synthesized to make its location in getEmittedMembers()
+  // deterministic.
```
Nice "why" comment!

davidungar commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -2417,13 +2418,64 @@ NamingPatternRequest::evaluate(Evaluator &evaluator, VarDecl *VD) const {
+  void add(AbstractFunctionDecl *afd) {
+    Mangle::ASTMangler mangler;
+    std::string mangledName;
+    if (auto *cd = dyn_cast<ConstructorDecl>(afd))
+      mangledName = mangler.mangleConstructorEntity(cd, /*allocator=*/false);
+    else if (auto *dd = dyn_cast<DestructorDecl>(afd))
+      mangledName = mangler.mangleDestructorEntity(dd, /*deallocating=*/false);
+    else
+      mangledName = mangler.mangleEntity(afd);
```
Too bad we don't just have a member function on AbstractFuncDecl to compute this "mangledNameForSorting". I wonder if this functionality is duplicated elsewhere?

slavapestov replied:
mangleEntity() will assert if you pass in a destructor or constructor. The reason we have separate entry points for those is that usually they have multiple manglings, for the (de)allocating and (de)initializing entry points, respectively. I don't think it's worth factoring this out.

davidungar reviewed (approved):
LGTM! You'll see I have teeny nit, but it's probably not worth the time to address it.

slavapestov committed:
ASTMangler: Fix assertion when mangling invalid identifiers

We sometimes need to mangle invalid identifiers, for example when sorting
declarations in EmittedMembersRequest. Make sure this doesn't assert.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility Debug

slavapestov commented:
@swift-ci Please smoke test

compnerd commented:
I think that this change might be the one responsible for a recent regression with `@main` usage (CC: @natecook1000):

```swift
// RUN: %target_swiftc -parse-as-library -parse-stdlib -module-name Swift -S -o - %s
@main
class S {
  public static func main() {
  }
}
```

The `-parse-as-library` is required for `@main` to be usable, `-parse-stdlib -module-name Swift` is purely to avoid the dependency on the standard library (trying to create a testcase for the future), You will note that there is a call emitted to `$ss1SC5$mainyyFZ` (on Linux x86_64, it will be `callq ($ss1SC5$mainyyFZ)@PLT` and on Windows x86_64 it will be `callq ($ss1SC5$mainyyFZ)`).  This method is not being synthesized any longer, resulting in undefined references.

lxbndr commented:
I have another sample code which not builds:
```
public class SomeClass {
}
extension SomeClass: Encodable {	
}
```

Linker tells about unresolved externals:
```
srx0-0f6207.o : error LNK2019: unresolved external symbol $s4srx09SomeClassC10CodingKeys33_0A88DACEEBAF8928C0B22F7967630AAFLLOs0D3KeyAAMc referenced in function $s4srx09SomeClassC10CodingKeys33_0A88DACEEBAF8928C0B22F7967630AAFLLOAFs0D3KeyAAWl
srx0.dll : fatal error LNK1120: 1 unresolved externals
```

 Noticed on Windows, but I guess it is multiplatform. I am pretty sure it is caused by this change. [Previous commit](https://github.com/apple/swift/commit/4b2fecf7580d6af3bfecf0bf15c41177f571b177) does not fail.

slavapestov commented:
@compnerd @lxbndr Thanks for the bug report, I'll take a look.
